Create database if not exists

AmitMY opened the pull request:
Adds ability to support CREATE DATABASE IF statement in CI_DB_Forge classes.
It defaults to false, as I am not sure of the usage in every single DB driver, but in MySQL & MySQLIi set to the correct statement. Other drivers does not support, but have the new signature, and class method documentation was added

This of course needs extending from people who know how to work with all DB drivers, and is just a start

narfbg commented:
That's not what I meant [here](http://forum.codeigniter.com/thread-66386-post-337546.html#pid337546).

I will **not** accept any patches that only implement MySQL.
Also, you're not complying with our [styleguide](https://codeigniter.com/userguide3/general/styleguide.html) (in fact, you're explicitly breaking it).

And you didn't need to close #4859 ... you're still pushing the same commits to the same branch.

AmitMY commented:
Ok, I will try to figure out all drivers format.
I was not aware about the thing with the pull request, so thanks

About the styleguide, I don't know where do I break it?
I just copied the same parameter from `create_table`, and the same comment
The only thing that was not copied from another place is:
```
$statement = ($if_not_exists === FALSE) ? '_create_database' : '_create_database_if';
if ($this->$statement === FALSE)
```
And I don't see nothing about the short 'if' operator `?:` in the style guide, or anything about an object dynamic access..

Sorry for not being fully professional, just trying to improve the current system for general needs

narfbg commented:
> About the styleguide, I don't know where do I break it?

I linked it so that you could read it ... if I pointed out every detail on the spot, nobody would read it. I know, we're all lazy, but you need to be aware of it.

>  I just copied the same parameter from `create_table`, and the same comment

You didn't *just* copy something, you have not only additions, but also modifications.
Also, I sincerely hope that your way of coding isn't just copy-pasting stuff around ...

> Sorry for not being fully professional, just trying to improve the current system for general needs

I myself treat it as a hobby, nobody expects full professionalism. Just a reasonable effort.

Portaflex commented:
"I will not accept", "Just a reasonable effort". Son, take a closer look to your attitude towards others. It does not seem to me much open and joined effort. Follow what Mr. Ezell does. And don't waste your time with an answer. Thanks and best regards.

narfbg commented:
You should try maintaining a project similar to CodeIgniter for a few years, on your own time, by yourself.

Absorb all the public pressure, countless questions from people too lazy to read a manual, feature requests for everything you could imagine (and then some), criticism (warranted or not), and all the hate that everything popular will naturally attract.
Put in the effort to fix a thousand bugs in code that you've "inherited", write manuals and guidelines (and watch people ignore them), review every single issue and pull request, assist people in debugging and adjusting their patches, etc.

And even then, "son", you should first learn about something called [context](https://github.com/bcit-ci/CodeIgniter/issues?q=author%3AAmitMY+created%3A%3C2016-10-20+sort%3Acreated-asc) before self-appointing yourself a member of the attitude police and try to tell me how to behave.
You know ... see if there's a reason why you're outraged, but the recipient of my "attitude" isn't.

I promise you, Mr. Ezell too will have a few of these moments in time.

AmitMY commented:
I agree that you can and should be harsh owner. However, I will say it is sometimes too much? 

For example:
> I linked it so that you could read it ... if I pointed out every detail on the spot, nobody would read it. I know, we're all lazy, but you need to be aware of it.

I read the styleguide prior to contributing, and read it again after you prompted me to. I did not find problems.

It is possible my head is too scatered and I can't understand where I was going off style, so I did not respond.

> You didn't just copy something, you have not only additions, but also modifications.

Yes, I did have modifications. I have modified function parameters, and fixed return signatures from "String" to "Bool"

> Also, I sincerely hope that your way of coding isn't just copy-pasting stuff around ...

Man, too harsh. All I meant is that I copied the same way you did "created table if not exists". meaning same PHPDoc comment for the parameter, same parameter name, and same documentation. (because it is the same)

>I myself treat it as a hobby, nobody expects full professionalism. Just a reasonable effort.

This was more than reasonable effort for me. I usually just open issues and expect the maintainers to resolve them, but here I saw it won't happen if I won't really push for it, so I tried a pull request (first time)

---

I wanted to write this at the time you answered, but realized I will be just wasting your time. I understand that you have here the same problem as SO - simple users are making things worse, while average+ users can make things better.

I accept my place as a simple user, and will (for now at least) use my own fork.

Again, sorry for wasting your time :)

narfbg commented:
> I agree that you can and should be harsh owner. However, I will say it is sometimes too much? 

Sometimes, maybe. But not in this case.
This was the 7th issue/pull request you opened in a week:

- #4849 - invalid
- #4850 - invalid; usage directly contradicting the manual; ignoring my request from #4839 to only report bugs here
- #4851 - your first valid bug report (though I should note - for a fairly contrived use case)
- #4852 - a somewhat contentious feature request, but ok still
- #4854 - invalid; based on not knowing the underlying technology; again ignoring my request from #4839 
- #4859 - an older version of this pull request, which you didn't have to close; I assume you just didn't know that you could modify already submitted patches, but you can't blame me for being annoyed at people using our repository as Git & GitHub learning playground

I did also reply to some of [your forum threads](http://forum.codeigniter.com/search.php?action=results&sid=4543024924437ae55a8a01972bc2bf6c) in the meantime. Not saying that there was anything wrong there, just further emphasizing on the unusual amount of attention that you required in a short timespan.

I understand you're new to this, and I do cut you some slack for that, but do you see how that may become annoying at some point?

And yet, now that I *had* to summarize all of this just to make my POV clear, someone will unquestionably find a way to paint it as me attacking you for no reason.
This is not an attack, and trust me - it's not even harsh, by my standards anyway.

> I read the styleguide prior to contributing, and read it again after you prompted me to. I did not find problems.

Well ... if I can spot the problems, and you can't find them after reading the styleguide a second time, isn't it logical that you may've *skimmed* through it as opposed to actually trying to understand it?

The *current* state of the patch no longer has that, but yet again just to make my point - [this commit](https://github.com/bcit-ci/CodeIgniter/pull/4860/commits/d3e37969712a1cde1a3ce12209935a6e818aaa21) was the culprit, and I was referring to you removing these spaces:

```diff
-    public function create_database($db_name = '')
+    public function create_database($db_name='', $if_not_exists = FALSE)
```

You're adding another parameter on the very same line using a different style yourself. It's almost as if those spaces are removed to be intentionally inconsistent ...
And no, I'm not saying you did it intentionally, I'm saying that's how it looks like because *if there was reasoning behind it*, that would be it.

(yes, I am nit-picking)

Nobody needs to read a styleguide to see such an inconsistency, so:

> It is possible my head is too scatered and I can't understand where I was going off style, so I did not respond.

No, it's a fork-in-the-eye-obvious thing.

>> You didn't just copy something, you have not only additions, but also modifications.
>
> Yes, I did have modifications. I have modified function parameters, and fixed return signatures from "String" to "Bool"

And I was pointing that out, because you had [previously](https://github.com/bcit-ci/CodeIgniter/pull/4860#issuecomment-255221104) said this:

> I just copied the same parameter from create_table, and the same comment
> The only thing that was not copied from another place is:
>
><pre>$statement = ($if_not_exists === FALSE) ? '_create_database' : '_create_database_if';
if ($this->$statement === FALSE)</pre>

... which is obviously a false claim, and more importantly - *you're not supposed* to make unrelated modificiations.
Styling changes / comment adjustments, general refactoring, bug fixes and new features don't belong together in the same patch.

I honestly don't see how *anybody* could be unhappy with me pointing this out. This is literally code reviews 101 material.

>> Also, I sincerely hope that your way of coding isn't just copy-pasting stuff around ...
>
> Man, too harsh. All I meant is that I copied the same way you did "created table if not exists". meaning same PHPDoc comment for the parameter, same parameter name, and same documentation. (because it is the same)

I know what you meant, and that was the problem. Not my own problem, but one that the programming world as a whole has - a culture of "copy-paste developers".

This is a topic too long to explain here, and granted - not a direct problem when you isolate this PR, but every contributor who says they've "copied" something from somewhere will get that response from me, because I've grown to hate it that much.

If that's harsh - so be it, but it's rather for your own good.

>> I myself treat it as a hobby, nobody expects full professionalism. Just a reasonable effort.
>
> This was more than reasonable effort for me. I usually just open issues and expect the maintainers to resolve them, but here I saw it won't happen if I won't really push for it, so I tried a pull request (first time)

If you've reported a valid bug, that's a reasonable expectation to have. But this isn't the case here.

Here, you're requesting that we merge your code, and you cannot reasonably expect that any serious project would do that if your code doesn't "fit". Because otherwise that project is doomed to fail.

And really, you've put it like I'm blaming you for submitting a pull request ... this is far from the case. Code contributions are naturally welcome, but also naturally require more effort.

Asking you to adjust your patch, remove unrelated (often unwanted) changes, to follow contribution guidelines, is a routine task for any OSS maintainer.

> I wanted to write this at the time you answered, but realized I will be just wasting your time. I understand that you have here the same problem as SO - simple users are making things worse, while average+ users can make things better.
>
> I accept my place as a simple user, and will (for now at least) use my own fork.

We do have a problem with people taking everything for granted and not investing the slightest of efforts - you're right on the money with that.
But it's not all black and white, and there's no classism angle to it.

Not that you'd ever see me categorize an individual as a "simple", "average" or "excellent" user, but if you've made the effort to submit a pull request - you're rather trying to make it better (as opposed to what you've said for "simple" users). The real problem is, that's only the first step and hence why I said "*reasonable* effort".

---

And this is to *everybody* reading, not just the people involved so far: *Stop assuming hostility!*

Unless you're a child, you can't expect everybody to sugar-coat their words for you. Especially when writing comments on the internet.

AmitMY commented:
Thanks for giving the time to write this! I finally know what was wrong.
I will not continue this pull request, because you want this to be supported in all DB drivers, support I cannot give, because of lack of testing environments for all DB drivers.

Closed
